util/project_allocation_mapper.py

```diff
@@ -0,0 +1,509 @@
+            self.set_attr(project, 'allocations', project_allocs)
+            return project
+        else:
+            tas_project = tas_proj(project_id)
+            try:
+                extras = ProjectExtras.objects.get(tas_project_id=tas_project.id)
+                tas_project.__dict__['nickname'] = extras.nickname
+            except ProjectExtras.DoesNotExist:
+                project_nickname = None
+            return tas_project
+
+    def allocation_approval(self, data, host):
+        if self.is_from_db:
+            # update allocation model
+            alloc = portal_alloc.objects.get(pk=data['id'])
+            data['status'] = data['status'].lower()
+            data['dateReviewed'] = datetime.now(pytz.utc)
+            for item in ['reviewerId', 'dateReviewed', 'start', 'end', 'status', 'decisionSummary', 'computeAllocated']:
+                setattr(alloc, allocation.TAS_TO_PORTAL_MAP[item], data[item])
+            alloc.save()
+            logger.info('Allocation model updated: data=%s', alloc.__dict__)
+            # send email to PI
+            email_args = {'charge_code': data['project'],
+                          'requestor_id': data['requestorId'],
+                          'status': data['status'],
+                          'decision_summary': data['decisionSummary'],
+                          'host': host}
+            self._send_allocation_decision_notification(**email_args)
+        else:
+            result = self.tas.allocation_approval(data['id'], data)
+            logger.info('Allocation approval TAS response: data=%s', json.dumps(result))
+            
+    def is_geni_user_on_chameleon_project(self, username):
+        on_chameleon_project = False
+        if self.is_from_db:
+            fed_proj = portal_proj.objects.filter(charge_code=settings.GENI_FEDERATION_PROJECTS['chameleon']['charge_code'])
+            keycloak_client = KeycloakClient()
+            for proj in keycloak_client.get_user_projects_by_username(username):
+                if proj.id == fed_proj.id:
+                    on_chameleon_project = True
+                    break
+        else:
+            fed_proj = tas_proj(settings.GENI_FEDERATION_PROJECTS['chameleon']['id'])
+            on_chameleon_project = any(u.username == username \
+                                        for u in fed_proj.get_users())
+        return on_chameleon_project
+    
+    def get_keystone_project_required_info(self, project):
+        if self.is_from_db:
+            return project.chargeCode, project.nickname, project.description
+        else:
+            project_extras = ProjectExtras.objects.filter(tas_project_id=project.id)
+            name = project.chargeCode
+            if project_extras:
+                name = project_extras[0].nickname
+            return project.chargeCode, name, project.description
+        
+    def add_user_to_project(self, tas_project, username):
+        if self.is_from_db:
```

Comment:
I think actually most of these calls need to check the existence of the project in Portal before doing anything.